File: backend/app/utils/renewal_frequency.py
"""Allowed renewal frequency values for insurance policies."""

from __future__ import annotations

RENEWAL_FREQUENCY_MONTHLY = "monthly"
RENEWAL_FREQUENCY_QUARTERLY = "quarterly"
RENEWAL_FREQUENCY_HALF_YEARLY = "half_yearly"
RENEWAL_FREQUENCY_YEARLY = "yearly"

RENEWAL_FREQUENCY_DEFAULT = RENEWAL_FREQUENCY_YEARLY

RENEWAL_FREQUENCY_VALUES = frozenset(
    {
        RENEWAL_FREQUENCY_MONTHLY,
        RENEWAL_FREQUENCY_QUARTERLY,
        RENEWAL_FREQUENCY_HALF_YEARLY,
        RENEWAL_FREQUENCY_YEARLY,
    }
)

RENEWAL_FREQUENCY_LABELS = {
    RENEWAL_FREQUENCY_MONTHLY: "Monthly",
    RENEWAL_FREQUENCY_QUARTERLY: "Quarterly",
    RENEWAL_FREQUENCY_HALF_YEARLY: "Half-Yearly",
    RENEWAL_FREQUENCY_YEARLY: "Yearly",
}


def validate_renewal_frequency(value: str | None) -> str:
    if value is None or not str(value).strip():
        raise ValueError("renewal frequency is required")
    normalized = str(value).strip().lower()
    if normalized not in RENEWAL_FREQUENCY_VALUES:
        raise ValueError("invalid renewal frequency")
    return normalized


def renewal_frequency_label(value: str | None) -> str:
    if value is None or not str(value).strip():
        return RENEWAL_FREQUENCY_LABELS[RENEWAL_FREQUENCY_DEFAULT]
    normalized = str(value).strip().lower()
    return RENEWAL_FREQUENCY_LABELS.get(normalized, value)

File: frontend/src/lib/insurance/renewal-frequency.ts
export const RENEWAL_FREQUENCY_MONTHLY = "monthly";
export const RENEWAL_FREQUENCY_QUARTERLY = "quarterly";
export const RENEWAL_FREQUENCY_HALF_YEARLY = "half_yearly";
export const RENEWAL_FREQUENCY_YEARLY = "yearly";

export const RENEWAL_FREQUENCY_DEFAULT = RENEWAL_FREQUENCY_YEARLY;

export type RenewalFrequencyValue =
  | typeof RENEWAL_FREQUENCY_MONTHLY
  | typeof RENEWAL_FREQUENCY_QUARTERLY
  | typeof RENEWAL_FREQUENCY_HALF_YEARLY
  | typeof RENEWAL_FREQUENCY_YEARLY;

export const RENEWAL_FREQUENCY_OPTIONS: Array<{ value: RenewalFrequencyValue; label: string }> = [
  { value: RENEWAL_FREQUENCY_MONTHLY, label: "Monthly" },
  { value: RENEWAL_FREQUENCY_QUARTERLY, label: "Quarterly" },
  { value: RENEWAL_FREQUENCY_HALF_YEARLY, label: "Half-Yearly" },
  { value: RENEWAL_FREQUENCY_YEARLY, label: "Yearly" },
];

export const RENEWAL_FREQUENCY_LABELS: Record<RenewalFrequencyValue, string> = {
  [RENEWAL_FREQUENCY_MONTHLY]: "Monthly",
  [RENEWAL_FREQUENCY_QUARTERLY]: "Quarterly",
  [RENEWAL_FREQUENCY_HALF_YEARLY]: "Half-Yearly",
  [RENEWAL_FREQUENCY_YEARLY]: "Yearly",
};

export function formatRenewalFrequencyLabel(value: string | null | undefined): string {
  if (!value || !value.trim()) {
    return RENEWAL_FREQUENCY_LABELS[RENEWAL_FREQUENCY_DEFAULT];
  }
  const normalized = value.trim().toLowerCase() as RenewalFrequencyValue;
  return RENEWAL_FREQUENCY_LABELS[normalized] ?? value;
}

export function validateRenewalFrequency(value: string): string | null {
  if (!value.trim()) {
    return "Please select a renewal frequency.";
  }
  const normalized = value.trim().toLowerCase();
  if (!RENEWAL_FREQUENCY_OPTIONS.some((option) => option.value === normalized)) {
    return "Please select a renewal frequency.";
  }
  return null;
}
